### ISSUE TITLE
Goals of backend plugin

### ISSUE THREAD

**Issue by markus2330** [ElektraInitiative/libelektra#3916]
> @kodebach wrote in https://github.com/ElektraInitiative/libelektra/issues/3912#issuecomment-873409568
> I really don't like that approach and one of the goals of #3693 is that plugins do not load other plugins. Only libelektra-kdb loads plugins and then plugins can get a reference to another plugin from libelektra-kdb.

Why would you want to have that goal? It sounds a bit like elektraMalloc (centralized resource management).

The original goal was backend=plugin, i.e. nested plugins.

**Comment by kodebach** [comment]
> The original goal was backend=plugin, i.e. nested plugins.

This is still the goal. A backend plugin is just a normal plugin that also fulfils a strict contract with `libelektra-kdb`. Nested plugins are also still the goal, in the sense of "one plugin calls another plugin".

However, plugins _loading_ other plugins directly is not a goal.

> centralized resource management

That is exactly the point. Although I would compare it more to `malloc` than to `elektraMalloc`. The idea is that `libelektra-kdb` knows how plugins should be loaded and can keep track of plugins more easily than, if any plugin could own the `Plugin *` to any other plugin.

If needed, we could still have the option that plugin A tells `libelektra-kdb` "please get me a new instance of plugin B". However, the central approach also makes it possible for plugn A to say "please give me a reference to the instance of plugin B for my mountpoint". In the current system, you need to manually ensure that e.g. only one instance of `type` is loaded. For example, `pluginA` and `pluginB` could both depend on `type` and load it dynamically. In the current system using the same `type` instance in to different plugins is pretty hard, unless the plugins know about each other or there is a "parent" plugin that loaded both `pluginA` and `pluginB` and provided them with the `type` instance.

In short, IMO the whole system because a lot easier, if plugin instances (i.e. loading and unloading) are managed by `libelektra-kdb`. When and how non-backend plugins are called, will be entirely up to the plugins. `libelektra-kdb` will only call the backend plugins according to the new contract.

**Comment by markus2330** [comment]
> That is exactly the point. Although I would compare it more to malloc than to elektraMalloc. The idea is that libelektra-kdb knows how plugins should be loaded and can keep track of plugins more easily than, if any plugin could own the Plugin * to any other plugin.

Might make sense depending on the details. Elektra's module loading system (or actually dlopen) already takes care of multiple loading of libs, so centralized plugin management only makes sense for returning a reference to an already loaded plugin (as you describe). But is this a task for libelektra-kdb or rather the backend plugin? Do we need references across several backends?

We should also consider to make one more plugin, let us say it is the "distributor" plugin, which parses the mountpoints and builds up all backend plugins. To separate this from `libelektra-kdb` has the advantage of making alternative implementation to Elektra (like Rust for @lawli3t) much more simple because then `libelektra-kdb` is basically a wrapper calling functions of the distributor plugin. @kodebach What do you think?

**Comment by kodebach** [comment]
> What do you think?

Short answer: I think we should stop discussing ideas for the backend plugin until we had a meeting and talked about the ideas and concepts in #3693. 

Long answer below. But first:

> the backend plugin

Please be clear in your terminology. I always make sure to talk about either:
1. "**a** backend plugin", "backend plugins" or "the [designated] backend plugin [for a mountpoint]": Any plugin that can be used as a the designated backend plugin for mountpoint. It must abide by the contract that backend plugins have with `libelektra-kdb`.
2. OR "**the `backend`** plugin": The default file-based implementation of a backend. The plugin is called `backend`.

---
**Long Answer:**

> But is this a task for libelektra-kdb or rather the backend plugin?

It is not a task for the `backend` plugin, because there could be other backend plugins (e.g. Database-based). We could mandate that all backend plugins, handle plugin management. However, this has several downsides:
1. We still need to define a standard API, so that all plugins are compatible with all backend plugins.
2. If something needs to change in the plugin management, all backend plugins must be updated. This could lead to inconsistencies.
3. In most cases, the implementation for this plugin management would be the same, so there would be unnecessary code duplication. Different implementations could also lead to bugs that are easily missed.

I think you already understood that and that's why you suggested the "distributor".

> Do we need references across several backends?

I don't know of a use case and I don't plan to support it.

> We should also consider to make one more plugin, let us say it is the "distributor" plugin, which parses the mountpoints and builds up all backend plugins. 

The current idea in #3693 already takes care of this (**):
1. `libelektra-kdb` parses the core part of the mountpoint definition. Specifically, it loads all configured plugins (defined in `system:/elektra/mountpoints/<mountpoint>/plugins/#`) with the configs (just loading, no actually mounting is done here). It then looks at `system:/elektra/mountpoints/<mountpoint>/backend` to find out, which of these plugins is designated as backend plugin for the mountpoint. This is the only plugin `libelektra-kdb` will interact with.
2. The rest of the mountpoint-definition is passed to the designated backend plugin, which will then actually assign the plugins to positions, check that required plugins types (e.g. storage, resolver) are present or whatever else might be required for this kind of backend (e.g. establish a database connection).

Part 1 could be extracted into a distributor "plugin" (*). But I don't think this would work well. We need to store the parsed mountpoints/backends somewhere. Currently this happens internally in `struct _KDB`. This gives use 2 options:
   1. The distributor "plugin" accesses `struct _KDB` directly. This would force any alternative `libelektra-kdb` to use the same `struct _KDB`. (A very bad idea)
   2. The distributor "plugin" uses its own internal data structure. This means we need more API to let `libelektra-kdb` access this internal data. (unnecessarily complicated)

> To separate this from `libelektra-kdb` has the advantage of making alternative implementation to Elektra [...] much more simple

I don't think this would be the case. In fact I think the distributor "plugin" would make things more complicated, because we would introduce another very strict API which alternative implementations have to conform to.

> because then `libelektra-kdb` is basically a wrapper calling functions of the distributor plugin.

Yes, if we take it to the extreme, where `libelektra-kdb` only calls functions from `distributor`, then an alternative implementation is very simple. However, the alternative implementation would loose all benefits, because it isn't doing any work anymore.

A final note on a Rust implementation specifically: AFAIK for Rust you can basically assume: the more things are written in Rust the easier it gets. Interfacing between Rust and C can be pretty hard and you can quickly loose lots of the benefits of Rust. Especially in terms of memory safety. Also, I think that the code for mountpoint parsing would be easier to write in Rust, so if anything C should reuse the Rust implementation.

----

(*) not really a plugin, since you can't swap it for something similar. All version need to parse the config exactly the same, because literally everything else depends on this. If anything, this could be separately linked library (e.g. `libelektra-plugins`) so an alternative implementation could re-use the code.

(**) If you want to take look, Part 1 is implemented [in this file](https://github.com/ElektraInitiative/libelektra/blob/898990819649541dafeca1b74717e4e1ba24e1e1/src/libs/elektra/kdb.c) in the functions `elektraMountpointsParse`, `parseAndAddMountpoint` and `openPlugins`. The code is a bit long, but that is mostly because it is C error handling and memory managment are complicated.
